invokeai/app/services/shared/invocation_context.py

```diff
@@ -426,10 +430,161 @@ def search_by_attrs(
+            subfolder or path. The following are acceptable alternatives:
+                stabilityai/stable-diffusion-v4
+                stabilityai/stable-diffusion-v4:fp16
+                stabilityai/stable-diffusion-v4:fp16:vae
+                stabilityai/stable-diffusion-v4::/checkpoints/sd4.safetensors
+                stabilityai/stable-diffusion-v4:onnx:vae
+
+            Because a local file path can look like a huggingface repo_id, the logic
+            first checks whether the path exists on disk, and if not, it is treated as
+            a parseable huggingface repo.
+
+        Returns:
+            Key to the newly installed model.
+
+        May Raise:
+            ValueError            -- bad source
+            UnknownModelException -- remote model not found
+            InvalidModelException -- what was retrieved from remote is not a model
+            TimeoutError          -- model could not be installed within timeout
+            Exception             -- another error condition
+        """
+        installer = self._services.model_manager.install
+        job = installer.heuristic_import(
+            source=source,
+            config=config,
+            access_token=access_token,
+            inplace=inplace,
+        )
+        installer.wait_for_job(job, timeout)
+        if job.errored:
+            raise Exception(job.error)
+        key: str = job.config_out.key
+        return key
+
+    def download_and_cache_ckpt(
+        self,
+        source: Union[str, AnyHttpUrl],
+        access_token: Optional[str] = None,
+        timeout: Optional[int] = 0,
+    ) -> Path:
+        """
+        Download the model file located at source to the models cache and return its Path.
+
+        This can be used to single-file install models and other resources of arbitrary types
+        which should not get registered with the database. If the model is already
+        installed, the cached path will be returned. Otherwise it will be downloaded.
+
+        Args:
+          source: A URL or a string that can be converted in one. Repo_ids
+                  do not work here.
+          access_token: Optional access token for restricted resources.
+          timeout: Wait up to the indicated number of seconds before timing
+                   out long downloads.
+
+        Result:
+          Path of the downloaded model
+
+        May Raise:
+          HTTPError
+          TimeoutError
+        """
+        installer = self._services.model_manager.install
+        path: Path = installer.download_and_cache(
+            source=source,
+            access_token=access_token,
+            timeout=timeout,
+        )
+        return path
+
+    def load_ckpt_from_url(
+        self,
+        source: Union[str, AnyHttpUrl],
+        access_token: Optional[str] = None,
+        timeout: Optional[int] = 0,
+        loader: Optional[Callable[[Path], Dict[str | int, Any]]] = None,
+    ) -> LoadedModel:
```

Comment:
`load_ckpt_from_url(...)` currently calls `download_and_cache_ckpt(...)` internally. What do you think of this alternative split?:
- `def download_and_cache_ckpt(self, source: Union[str, AnyHttpUrl]) -> Path`
    - No change from what it does now.
- `def load_custom_model(self, path: Path) -> LoadedModel`
    - Loads a model from a path and registers it with the model cache.
    - Raises if the path does not exist.

This gives better separation of concerns and gives more flexibility to the user. For example, it gives the ability for a community node to load custom models from disk without needing them to be publicly downloadable.

The one downside I can think of is that it would cause an extra check that the model file exists on disk in cases where the model is already in the cache. Maybe this is why you did it the other way to begin with? This is definitely solvable through caching if it actually begins a performance issue.

Reply:
The split is fine with me. How about we keep `load_ckpt_from_url()` as a three-liner that does:

```
model_path = download_and_cache_ckpt()
load_custom_model(model_path)
```

There is no need for an additional path exists check, as `download_and_cache_ckpt()` already does this, and will raise an exception if the file neither exists in the cache nor can be downloaded.

I haven't implemented this yet - let me know what you think.

Reply:
> The split is fine with me. How about we keep load_ckpt_from_url() as a two-liner that does:

Sure, that sounds good to me.

> There is no need for an additional path exists check, as download_and_cache_ckpt() already does this, and will raise an exception if the file neither exists in the cache nor can be downloaded.

Right. What I meant was that by splitting it up, we end up doing this check every time. Whereas with the current `load_ckpt_from_url(...)` it can be skipped if the model is in the cache.

Reply:
Ok, done.